### PR TITLE
EOS-20379: Health graph implementation

### PR DESCRIPTION
**Pull request by Vrishali-web-dev**
Signed-off-by: Vrishali Danave <vrishali.danave@seagate.com>

# Frontend

# Problem Statement

- _Overview: JIRA number/GitHub Issue added to PR: [Y/N]:_
- _Describe the problem this patch intends to solve._

## Design

- _For Bug describe the fix here._
- _For feature, post the link to the solution page._
- _Any northbound interface change and has been notified to other gatekeepers [Y/N]:_

## Coding

- _Coding conventions are followed and code is consistent. Yes/No?_
- _Confirm All CODACY errors are resolved. Yes/No?_

## Testing

- _Confirm that Test Cases are added (for both the cases, fix and feature). Yes/No?_
- _Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability. Yes/No?_
- _Confirm Testing was performed with installed RPM. Yes/No?_
- _Confirm all the ut/regression/smoke test has been performed [Y/N]_

## Checklist

- _PR is self-reviewed? Yes/No?_
- _GitHub Issue is updated_
- _Jira is updated_
  -  _Check if the description is clear and explained._ 
    -  _Check Acceptance Criterion is defined._
      -  _All the tests performed should be mentioned before Resolving a JIRA._
        -  _Verification needs to be done before marked as Closed/Verified_
        - _Any interface change Yes/No?_
        - _Change notified to other gatekeepers: Yes/No?_
        - _Code reviews done and addressed: Yes/No?_
        - _Codacy issues reviewed and addressed: Yes/No?_
        - _Deployment test : Done/Not?_
        - _UT/smoke test: Done/Not?_
        - _Jira number added to PR: Yes/No?_
        - _Github merge done using UI: Yes/No?_
        - _Side effects on other features - deployment/update/node-replacement_
        - _Changes to quick-start-guide/community impact_
        - _All dependent component code PR are raised or already merged Yes/No_
        - _All dependent code already merged in landing branch Yes/No_

#Solution/Screenshots:
Verified with mock data to show nodes of all the status

image

Screenshot with API response data:

image

### CODE REVIEW

File: gui/src/components/health/cortx-health-graphical.vue [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU Affero General Public License for more details.
+* You should have received a copy of the GNU Affero General Public License
+* along with this program. If not, see <https://www.gnu.org/licenses/>.
+* For any questions about this software or licensing,
+* please email opensource@seagate.com or cortx-questions@seagate.com.
+*/
+<template>
+  <div>
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:18]: Outer div not required unless we have multiple elements.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+    <div id="treeContainer"></div>
+  </div>
+</template>
+<script lang="ts">
+import CortxTabs, { TabsInfo } from "../widgets/cortx-tabs.vue";
+import { Component, Vue, Prop } from "vue-property-decorator";
+import { Api } from "../../services/api";
+import apiRegister from "../../services/api-register";
+import * as d3 from "d3";
+// import { tree } from "vued3tree";
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:28]: Please remove unused and commented code.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+import * as moment from "moment";
+
+@Component({
+  name: "cortx-health-graphical"
+})
+export default class CortxHealthGraphical extends Vue {
+  public healthQueryParams: any = {};
+  public clusterHealthData: any = {};
+
+  public mounted() {
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:38]: Please use async/await

File: gui/src/components/health/cortx-health-graphical.vue [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+import apiRegister from "../../services/api-register";
+import * as d3 from "d3";
+// import { tree } from "vued3tree";
+import * as moment from "moment";
+
+@Component({
+  name: "cortx-health-graphical"
+})
+export default class CortxHealthGraphical extends Vue {
+  public healthQueryParams: any = {};
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:35]: Is this variable required to be declared global? If not please declare it locally.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+      // Compute the new tree layout.
+      let nodes = treeData.descendants(),
+        links = treeData.descendants().slice(1);
+      // Normalize for fixed-depth.
+      nodes.forEach(function(d: any) {
+        d.y = d.depth * 180;
+      });
+      // ****************** Nodes section ***************************
+      // Update the nodes...
+      let i = 0;
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:128]: Is this variable used anywhere? If not please remove it.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+    // moves the 'group' element to the top left margin
+    const svg = d3
+      .select("#treeContainer")
+      .append("svg")
+      .attr("width", width + margin.right + margin.left)
+      .attr("height", height + margin.top + margin.bottom)
+      .style("overflow", "auto")
+      .append("g")
+      .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
+    const i = 0,
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:93]: Is this variable used anywhere? If not please remove it.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+      let node = svg.selectAll("g.node").data(nodes, function(d: any) {
+        return d.id || (d.id = ++i);
+      });
+      // Enter any new modes at the parent's previous position.
+      let nodeEnter = node
+        .enter()
+        .append("g")
+        .attr("class", "node")
+        .attr("transform", function(d: any) {
+          return "translate(" + source.y0 + "," + source.x0 + ")";
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:138]: you can use template literal here.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+        });
+      nodeEnter
+        .append("rect")
+        .attr("width", "140")
+        .attr("height", "100")
+        .attr("rx", 5)
+        .style("fill", "#fff")
+        .style("stroke", function(d: any) {
+          let color;
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:147]: we can create a function expression for this color logic as it is being used at multiple places.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+        .attr("x", "10px")
+        .attr("y", "60px")
+        .attr("id", function(d: any) {
+          return "nodeInfoTextID" + d.id + d.depth;
+        })
+        .style("visibility", "hidden")
+        .append("tspan")
+        .style("font-size", "10px")
+        .text(function(d: any) {
+          return "Name: " + d.data.resource;
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:242]: Please use template literals where you have string concatination.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+        })
+        .append("tspan")
+        .style("font-size", "10px")
+        .attr("x", "10px")
+        .attr("dy", "15px")
+        .text(function(d: any) {
+          return "ID: " + d.data.id;
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:249]: Please use template literals where you have string concatination.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+        .style("cursor", "pointer")
+        .on("click", click);
+      // UPDATE
+      let nodeUpdate = nodeEnter.merge(node);
+      // Transition to the proper position for the node
+      nodeUpdate
+        .transition()
+        .duration(duration)
+        .attr("transform", function(d: any) {
+          return "translate(" + d.y + "," + d.x + ")";
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:451]: Please use template literal.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+        });
+      // Remove any exiting nodes
+      let nodeExit = node
+        .exit()
+        .transition()
+        .duration(duration)
+        .attr("transform", function(d: any) {
+          return "translate(" + source.y + "," + source.x + ")";
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:459]: Please use template literal.

File: gui/src/components/health/cortx-health-graphical.vue [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+        .remove();
+
+      // Store the old positions for transition.
+      nodes.forEach(function(d: any) {
+        d.x0 = d.x;
+        d.y0 = d.y;
+      });
+      // Creates a curved (diagonal) path from parent to the child nodes
+      function diagonal(s: any, d: any) {
+        let path;
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:514]: this should be const as it is not being reassigned. also you can simply return template literal

File: gui/src/components/health/cortx-health-graphical.vue [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+        .append("image")
+        .attr("class", "node")
+        .attr("xlink:href", require("@/assets/info-icon.svg/"))
+        .attr("x", "115px")
+        .attr("y", "10px")
+        .attr("width", "20px")
+        .attr("height", "20px")
+        .style("cursor", "pointer")
+        .on("click", function(d: any) {
+          if (
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:270]: You can get all the required elements before if and simply set the attributes depending on condition

File: gui/src/components/health/cortx-health-graphical.vue [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+          for (let i = 0; i < obj.children.length; i++) {
+            traverse(obj.children[i]);
+          }
+        }
+      } else {
+        rename(obj);
+      }
+    }
+
+    function rename(obj) {
```
Comment (shrimetta) [inline gui/src/components/health/cortx-health-graphical.vue:65]: Need to check if we need to do this

**Review by shrimetta — CHANGES_REQUESTED** [review]
Please address the comments.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 5 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue:128]: Yes it's required.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 6 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue:93]: Removed this as not used anywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 15 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue:65]: Yes we need this function. We are updating key in object..

File: gui/src/components/health/cortx-health-graphical.vue (hunk 1 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Removed outer div

File: gui/src/components/health/cortx-health-graphical.vue (hunk 2 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Removed unwanted commented code

File: gui/src/components/health/cortx-health-graphical.vue (hunk 3 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Added async and await

File: gui/src/components/health/cortx-health-graphical.vue (hunk 4 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: removed variable from global scope and added it locally

File: gui/src/components/health/cortx-health-graphical.vue (hunk 7 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Updated to template literal everywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 8 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Created new function for colors.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 9 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Updated to template literal everywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 10 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Updated to template literal everywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 11 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Updated to template literal everywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 12 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Updated to template literal everywhere in a code.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 13 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Variable updated to const.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 14 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Created variables for elements.

File: gui/src/components/health/cortx-health-graphical.vue (hunk 15 above)
Comment (Vrishali-web-dev) [reply gui/src/components/health/cortx-health-graphical.vue]: Removed these functions and updated code to achieve the same results.

**Review by shrimetta — APPROVED** [review]

**Pull request merged**